### PR TITLE
Add support for yet another TI embedding format (main version)

### PR DESCRIPTION
**Pull request by lstein**
- This PR adds support for embedding files that contain a single key "emb_params". The only example I know of this format is the "EasyNegative" embedding on HuggingFace, but there are certainly others.

- This PR also adds support for loading embedding files that have been saved in safetensors format.

- It also cleans up the code so that the logic of probing for and selecting the right format parser is clear.

- This is the same as #3045, which is on the 2.3 branch.

### CODE REVIEW

File: invokeai/backend/stable_diffusion/textual_inversion_manager.py [hunk 1]
```diff
@@ -309,92 +295,72 @@ def _get_or_create_token_id_and_assign_embedding(
...
-        elif file_type == "bin":
-            return self._parse_embedding_bin(embedding_file)
-        else:
-            print(f"** Notice: unrecognized embedding file format: {embedding_file}")
+    def _parse_embedding(self, embedding_file: str)->dict:
+        suffix = Path(embedding_file).suffix
+        try:
+            if suffix in [".pt",".ckpt",".bin"]:
+                scan_result = scan_file_path(embedding_file)
+                if scan_result.infected_files == 1:
```
Comment (JPPhoto) [inline invokeai/backend/stable_diffusion/textual_inversion_manager.py:303]: Should we be testing > 0?

**Review by JPPhoto — APPROVED** [review]

**Review by damian0815 — CHANGES_REQUESTED** [review]

File: invokeai/backend/stable_diffusion/textual_inversion_manager.py [hunk 2]
```diff
@@ -309,92 +295,72 @@ def _get_or_create_token_id_and_assign_embedding(
...
+        embedding_info["name"] = embedding_ckpt["name"]
+                
+        # Check num of embeddings and warn user only the first will be used
+        embedding_info["num_of_embeddings"] = len(
+            embedding_ckpt["string_to_token"]
+        )
+        if embedding_info["num_of_embeddings"] > 1:
+            print("   | More than 1 embedding found. Will use the first one")
+        embedding = list(embedding_ckpt["string_to_param"].values())[0]
+        embedding_info["embedding"] = embedding
```
[the remark this answers is not in the document]
Comment (lstein) [reply invokeai/backend/stable_diffusion/textual_inversion_manager.py:344]: Thanks. I’ll implement option 1.

File: invokeai/backend/stable_diffusion/textual_inversion_manager.py [hunk 3]
```diff
@@ -413,7 +379,46 @@ def _handle_broken_pt_variants(
...
+        short_path = Path(filepath).parents[0].name+'/'+Path(filepath).name
+        
+        print(f'   | Loading v4 embedding file: {short_path}')
+        embedding_info = {}
+        if list(embedding_ckpt.keys()) == 0:
+            print(f"   ** Invalid embeddings file: {short_path}")
+            embedding_info = None
+        else:
+            for token in list(embedding_ckpt.keys()):
+                embedding_info["name"] = (
```
[the remark this answers is not in the document]
Comment (lstein) [reply invokeai/backend/stable_diffusion/textual_inversion_manager.py:417]: True. This seems to be a long-standing bug.  We don’t deal well with embeddings that have multiple tokens. I’ll have to change the upstream architecture to receive a list of the parsed dicts.

File: invokeai/backend/stable_diffusion/textual_inversion_manager.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lstein) [reply invokeai/backend/stable_diffusion/textual_inversion_manager.py]: Implemented using a dataclass to hold parameters - much easier to read now and less redundant.

File: invokeai/backend/stable_diffusion/textual_inversion_manager.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lstein) [reply invokeai/backend/stable_diffusion/textual_inversion_manager.py]: Fixed, although I have yet to see an embedding file with multiple tokens in the wild.

**Review by damian0815 — APPROVED** [review]
looks good - with a small nitpick, having read the last version of `textual_inversion_manager.py` i think `EmbeddingInfo` would better be named `ParsedEmbeddingInfo` and also its definition ought to go beneath `TextualInversionInfo` - or better, it ought to be a subclass of `TextualInversionManager` itself (can't recall if that works in python). otherwise its scope/purpose is a little unclear.

File: invokeai/backend/stable_diffusion/textual_inversion_manager.py (hunk 1 above)
Comment (lstein) [reply invokeai/backend/stable_diffusion/textual_inversion_manager.py]: yes. fixed

**Pull request merged**